src/formats/jpeg.zig

```diff
@@ -0,0 +1,1084 @@
+        var k: usize = 0;
+        var code: usize = 0;
+        var si = values[0].len;
+
+        while (values[k].len != 0) {
+            while (values[k].len == si) {
+                // huffcode[k] = code;
+                code += 1;
+                k += 1;
+            }
+
+            while (values[k].len != si) {
+                code <<= 1;
+                si += 1;
+            }
+        }
+    }
+
+    pub fn init(allocator: *Allocator, table_class: u8, reader: ImageReader) !HuffmanTable {
+        if (table_class & 1 != table_class)
+            return error.InvalidHuffmanTableClass;
+
+        var code_counts: [16]u8 = undefined;
+        if ((try reader.read(code_counts[0..])) < 16) {
+            return error.IncompleteHuffmanTable;
+        }
+
+        if (JPEG_DEBUG) std.debug.print("  Code counts: {any}\n", .{code_counts});
+
+        var total_huffman_codes: usize = 0;
+        for (code_counts) |count| total_huffman_codes += count;
+
+        var huffman_code_map = HuffmanCodeMap.init(allocator);
+        errdefer huffman_code_map.deinit();
+
+        if (JPEG_VERY_DEBUG) std.debug.print("  Decoded huffman codes map:\n", .{});
+
+        var code: u16 = 0;
+        for (code_counts) |count, i| {
+            if (JPEG_VERY_DEBUG) {
+                std.debug.print("    Length {}: ", .{i + 1});
+                if (count == 0) {
+                    std.debug.print("(none)\n", .{});
+                } else {
+                    std.debug.print("\n", .{});
+                }
+            }
+
+            var j: usize = 0;
+            while (j < count) : (j += 1) {
+                // Check if we hit all 1s, i.e. 111111 for i == 6, which is an invalid value
+                if (code == (@intCast(u17, 1) << (@intCast(u5, i) + 1)) - 1) {
+                    return error.InvalidHuffmanTable;
+                }
+
+                const byte = try reader.readByte();
+                try huffman_code_map.put(.{ .length_minus_one = @intCast(u4, i), .code = code }, byte);
+                code += 1;
+
+                if (JPEG_VERY_DEBUG) std.debug.print("      {b} => 0x{X}\n", .{ code, byte });
+            }
+
+            code <<= 1;
+        }
+
+        return HuffmanTable{
+            .allocator = allocator,
+            .code_counts = code_counts,
+            .code_map = huffman_code_map,
+            .table_class = table_class,
+        };
+    }
+
+    pub fn deinit(self: *HuffmanTable) void {
+        self.code_map.deinit();
+    }
+};
+
+const HuffmanReader = struct {
+    table: ?*const HuffmanTable = null,
+    reader: ImageReader,
+    byte_buffer: u8 = 0,
+    bits_left: u4 = 0,
+    last_byte_was_ff: bool = false,
+
+    pub fn init(reader: ImageReader) HuffmanReader {
+        return .{
+            .reader = reader,
+        };
+    }
+
+    pub fn setHuffmanTable(self: *HuffmanReader, table: *const HuffmanTable) void {
+        self.table = table;
+    }
+
+    fn readBit(self: *HuffmanReader) !u1 {
```

Comment:
There is `std.io.BitReader` that implement bit reading on top on a reader stream

Reply:
Can't believe I haven't seen this, will replace all this code with that after finishing up the other review comments :P

Reply:
Ah, on second thought, this can't be using `std.io.BitReader` because that doesn't support skipping stuffed bytes. This causes issues with images which contain such bytes. So for now I'm for keeping this logic.